### DISCUSSION TITLE
Using OnFrameMetricsAvailableListener API on SurfaceView

### DISCUSSION THREAD

**Discussion by mbalajee** [google/filament#7810] [Q&A]
This is not directly related to Filament but something I have to use because am using Filament. Cross posting here (originally posted in [reddit](https://www.reddit.com/r/androiddev/comments/1cggxec/using_onframemetricsavailablelistener_api_on/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button)).  Hoping to get some help here 🙏 

I need to use the new FrameMetric [API](https://developer.android.com/reference/android/view/Window#addOnFrameMetricsAvailableListener(android.view.Window.OnFrameMetricsAvailableListener,%20android.os.Handler)) introduced in Android N on the SurfaceView. Since that API is available only on the window property I can't use it on the SurfaceView (which has its own Window but not exposed unlike Activity#window)

I tried `activity.window.addOnFrameMetricsAvailableListener` on the activity to which the SurfaceView is added but the `onFrameMetricsAvailable` is never called. It gets called on activities without the SurfaceView (occupying whole layout)

So I believe I should call `addOnFrameMetricsAvailableListener` on the SurfaceView's window to get the callback

Why not use `Choreographer#FrameCallback#doFrame` to manually calculate the frames per second (just by having a counter, incrementing in doFrame & resetting every second)?
This fantastic video https://youtu.be/R6kqZQusv1U?t=517 explains why FPS is not that good of a metric & we should rather depend on FrameMetric for a reliable frame rate performance

Filament also accepts TextureView which shares the activity's window, so I think the callback should work just fine with TextView but it looks like SurfaceView is the recommended choice for 3D rendering. So we're sticking with SurfaceView

**Comment by romainguy** [comment]
What are you trying to do? Frame metrics can't come from a `SurfaceView` at the moment.

**Reply by mbalajee** [reply]
We're loading models (`.glb`) with node count varying from few tens to 62K (which is currently failing to load with [this](https://github.com/google/filament/discussions/7567) error on all platforms).  Filament can handle up to 12K nodes on most devices.  Filament integration is still in staging environment & will mostly land in production in the next quarter.  So we want to get as much metric info as possible in production for different types of devices so we can choose to optimize (like disabling postProcessing / hiding certain no of nodes from C++) for certain devices.

Currently I am getting the FPS info by incrementing a counter in Choreographer#doFrame callback.  Below is a part of the code,
```kotlin
   private val frameCallback = object : FrameCallback {
        override fun doFrame(frameTimeNanos: Long) {
            doFrameCallback.doFrame(frameTimeNanos)
            frameCount++
            choreographer.postFrameCallback(this)
        }
    }

    private fun startMonitoringFPS() {
        stopMonitoringFPS()
        fpsJob = lifecycle.coroutineScope.tryLaunch(logError = false) {
            while (lifecycle.currentState == Lifecycle.State.RESUMED) {
                delay(1000)
                val count = frameCount
                frameCount = 0
                fpsCallback?.fps(count)
            }
        }
    }
```

But after watching your [video](https://youtu.be/R6kqZQusv1U?t=517) I learnt that FPS is not a good metric and decided to use FrameMetric API info to get metrics info at various percentiles.  I basically want to get reliable frame rate information at runtime so we don't end up optimizing Filament based on false positives
